TAV-876: Node 20

passcod opened the pull request:
### Changes

- Node 16 is EOL
- Upgrade to Node 20 (current LTS)!
- Upgrade version of `pdfjs-dist`, as its dependency `canvas` needs to be upgraded to build with Node 20
- Upgrade to Electron 27
- Upgrade to `actions/checkout@v4` in actions ([uses Node.js 20](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20), though not related to Tamanu's usage)
- Use `.json5` extension for config files instead of `.json`, to avoid [a bug in `config`](https://github.com/node-config/node-config/issues/715)
- Upgrade all transitive dependencies to opt them into higher node version support.
- ⚠️ We use Node 18 for one step of building the desktop client in Docker/CI! In practice this shouldn't matter much but it might restrict us a bit until Tamanu Web lands.

### Checklist

- [x] Code is finished
- [x] Tests pass
- [x] Servers work
- [x] Desktop works

_Upon merging:_

- [x] Update the changelog (find it [here](https://github.com/beyondessential/tamanu/releases) in the appropriate _draft_ release)
  - [x] with a ticket reference (e.g. `TAN-123: a one line description`, keep the lists sorted)
  - [x] manual deployment steps
- [ ] Update the [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c) as needed
- [x] Add an issue due for **late 2025** to upgrade past Node 20 (one year before Node 20 goes out of maintenance, to give ourselves time to do it before it's six months until EOL)

passcod committed:
Node 20?

passcod committed:
Use checkout v4

passcod committed:
Docker Node 20

passcod committed:
Docker desktop node 20

passcod committed:
Upgrade bcrypt and dedup its versions

passcod committed:
More 16->20

passcod committed:
Use json5 extension for config files

passcod committed:
More default.json rewrites

passcod committed:
Update ci

passcod committed:
Remove desktop build

passcod committed:
Load json5 in webpack

passcod committed:
Jest ignore it

passcod committed:
why ain't there just one

dannash100 committed:
notam-49: does this work

dannash100 committed:
notam-49: it works

dannash100 committed:
notam-49: fix sync test

dannash100 committed:
notam-49: try lan test fix 2

dannash100 committed:
notam-49: template each tests

passcod committed:
Update mobile lock file

passcod committed:
Merge pull request #4918 from beyondessential/fix/notam-49-fix-tests

passcod committed:
Update lockfile

passcod committed:
Dual-fist versions

passcod committed:
Build clients as node 16

passcod committed:
tsconfig backward compat says what

passcod committed:
okay babel whatever you say

passcod committed:
One node version spec

passcod committed:
Maybe we can do 18

passcod committed:
Dual-fist node 20 and 18 (for desktop)

passcod committed:
Fix desktop's babel config

passcod committed:
i haaaaaate them

passcod committed:
oh yep restore that too

mcccclean reviewed (changes requested):
Looking good! Keen to get this merged ASAP to minimise conflicts showing up - I just had a couple questions about how things work that're probably better answered by leaving comments rather than telling me directly.

passcod committed:
Update babel.config.js

passcod committed:
Update Dockerfile

passcod committed:
fix desktop renderer

passcod committed:
update electron instead of polyfilling

passcod committed:
Revert "update electron instead of polyfilling"

This reverts commit f4a17fbeb12d081264f2e2fcb9a8522e24cec0c5.

There's incompatibilities from Electron upgrade.

passcod committed:
Pin buggy react-idle-timer

The value for the promptBeforeIdle property must be less than the timeout property

that was introduced post 5.4.2 and breaks our usage (as well as other people´s...)

passcod committed:
Revert "Revert "update electron instead of polyfilling""

This reverts commit f8b7834d01adf0fe309525ae32aa49f77951beda.

passcod committed:
@electron/remote

passcod committed:
Remote use of remote

passcod committed:
Replace getLocale

this will also work better for Tamanu Web

passcod committed:
Remove electron/remote dep

passcod committed:
Fix undefined JSON bug

passcod committed:
Merge branch 'main' into feature/notam-49-node-20

passcod committed:
Stop using chalk for silly reasons

passcod committed:
Re-add @electron/remote

passcod committed:
Unlint

passcod committed:
Get back fileMock

passcod committed:
Fix desktop tests againi

passcod committed:
Only run tests that are tests

passcod committed:
globalThis is a global AS PER THE ES SPEC YOU IDIOT ESLINT
